polling_stations/apps/data_finder/helpers.py

```diff
@@ -7,99 +7,222 @@
+    def geocode(self):
+        addresses = Address.objects.filter(postcode=self.postcode)
+        if not addresses:
+            raise ObjectDoesNotExist('No addresses found for postcode %s' % (self.postcode))
+
+        codes = self.get_codes(self.get_uprns(addresses))
+        centre = centre_from_points_qs(addresses)
+        return {
+            'source': 'addressbase',
+            'wgs84_lon': centre.x,
+            'wgs84_lat': centre.y,
+            'council_gss': codes['council_gss'],
+            'gss_codes': codes['gss_codes'],
+        }
 
-        try:
-            # attempt to parse error from json
-            res_json = res.json()
-            if 'error' in res_json:
-                raise PostcodeError("Mapit error {}: {}".format(res_json['code'], res_json['error']))
-            else:
-                raise PostcodeError("Mapit error {}: unknown".format(res.status_code))
-        except ValueError:
-            # if we fail to parse json, raise a less specific exception
-            raise PostcodeError("Mapit error {}: unknown".format(res.status_code))
+    def geocode_point_only(self):
+        addresses = Address.objects.filter(postcode=self.postcode)
+        if not addresses:
+            raise ObjectDoesNotExist('No addresses found for postcode %s' % (self.postcode))
 
-    res_json = res.json()
+        centre = centre_from_points_qs(addresses)
+        return {
+            'source': 'addressbase',
+            'wgs84_lon': centre.x,
+            'wgs84_lat': centre.y,
+        }
+
+
+def geocode_point_only(postcode, sleep=True, mapit=mapit_wrapper_factory):
+    addressbase = AddressBaseWrapper(postcode)
+    mapit = mapit(postcode)
+
+    try:
+        # first try addressbase
+        result = addressbase.geocode_point_only()
+    except ObjectDoesNotExist:
+        # we couldn't find this postcode in AddressBase: fall back to mapit
+
+        # optional sleep to avoid hammering mapit
+        if sleep:
+            time.sleep(1.3)
+
+        result = mapit.geocode()
+    except:
+        # something else went wrong: lets give mapit a go anyway
+
+        # optional sleep to avoid hammering mapit
+        if sleep:
+            time.sleep(1.3)
 
-    gss_codes = []
-    council_gss = None
-    for area in res_json['areas']:
-        if 'gss' in res_json['areas'][area]['codes']:
-            gss = res_json['areas'][area]['codes']['gss']
-            gss_codes.append(gss)
-            if res_json['areas'][area]['type'] in COUNCIL_TYPES:
-                council_gss = gss
-
-    if 'wgs84_lon' not in res_json or 'wgs84_lat' not in res_json:
-        raise PostcodeError("No location information")
-
-    return {
-        'wgs84_lon': res_json['wgs84_lon'],
-        'wgs84_lat': res_json['wgs84_lat'],
-        'gss_codes': gss_codes,
-        'council_gss': council_gss,
-    }
+        result = mapit.geocode()
+
+    return result
+
+
+def geocode(postcode, mapit=mapit_wrapper_factory):
```

Comment:
I reckon on 1. using [`mock`](https://docs.python.org/dev/library/unittest.mock.html) would be better. Something like:
```python
from mock import patch
class MyTest(unittest.TestCase):

    @patch('MapitWrapper.geocode')    
    def test_no_records(self, mocked_mapit_wrapper_geocode):
        mocked_mapit_wrapper_geocode.return_value = { 'source': 'mapit' }
        self.assertEqual('mapit', result['source'])
```

(Untested, but you get the idea)